Add module base class and move table tents to module

lostcarpark opened the pull request:
Made the Tools class into an abstract class that each module class can be inherited from. I think this will allow functionality to be added in a standard way that modules will be able to implement as required. As an example, I would like to add a cron function that will allow modules to run background tasks in a standard way.

I've also tweaked the module discovery process to allow modules to be placed in their own folder, based on namespace.

Finally, I've converted the Table Tents page into a module to make it better fit in the tools page. I did think about making this a separate commit, but it was difficult to untangle them.

bcholmes commented:
Hm. I'm a bit uncertain about the inheritance. I think a module is a thing that "has" (optionally) tools, but also (potentially, in the future) other things like menu options and other stuff. From an OO perspective, I don't think a module IS A tool; I think it HAS A (or HAS SOME) tools. 

Let me roll this around in my head a bit.

lostcarpark commented:
I see your point.

I think Having a base class for modules is important, but I think you're right that it shouldn't be Tool. I hadn't thought of the case where a module might have more than one Tool, but it certainly seems something we should allow for.

I think there will be cases where tools might want to use a subclass of Tools, but they probably shouldn't be forced to.

I'll do some reworking on this.

bcholmes commented on webpages/api/admin/module_model.php:
```diff
@@ -26,32 +26,56 @@ public function getDescriptorFileName() {
+            return 'PlanZ\Module\\';
+        }
+        // Split package prefix into words, and capitalize each word.
+        return 'PlanZ\\' . $this->makePascalCase($name) . '\\';
```
I think this represents a change in behaviour. I have some WisCon-specific modules that use "Wiscon" as the namespace; I wouldn't have expected "PlanZ" to be part of the namespace.

lostcarpark replied:
Thanks. I hadn't considered modules from outside the PlanZ project. I will look at this again.
Can you give an example of a package name and the resulting namespace for one of the WisCon modules?

bcholmes replied:
Sure. I have a module that creates the JSON extract for the WisSched iOS and Android apps (kinda-sorta comparable to what you do with ConClàr). 

The database package name is "wiscon.wissched", and the module definition is in module/wiscon/wissched_module.php with a namespace of "Wiscon".

lostcarpark replied:
I've revised, so it should now create the same namespace name.

The case where the namespace name would not be the same is where the first part of the package name contains underscores.

For example, `table_tents` would previously have produced the namespace `Table_tents`, but now produces `TableTents`.

lostcarpark commented:
I have added a new "ModuleBase" class in the "module" directory. It's intended that this can be the base class for module classes, but module classes that don't inherit it should work fine.

Added "PlanZ" namespace to some classes.

Changed the namespace generation to more closely replicate the old functionality, except where underscores are concerned.

Fixed some issues with base path that I'd missed.

lostcarpark commented:
Oh, the one thing I didn't think of till now is you'd need to add `use PlanZ\Tool;` to your WisCon modules. Would you be okay with that?

bcholmes commented on webpages/api/admin/module_model.php:
```diff
@@ -64,30 +96,30 @@ public function getDescriptorClass() {
+            SET is_enabled = ?
             WHERE id = ?;
-EOD;
+        EOD;
```
I think Leanne originally told me to keep the EOD aligned to the left because of reasons (otherwise it was a problem for some version of... um.. PHP? Or something).

lostcarpark replied:
Yes, in PHP < 7.3, the closing identifier must be left aligned. 

In PHP >= 7.3 the closing identifier can be indented, but must not be indented further than any line of text in the string. The same number of leading spaces get removed from each line of the string.

If we are allowing arrow functions, which require PHP 7.4, then indenting closing string identifiers aren't an issue.

bcholmes replied:
Okay. Good to know.

lostcarpark replied:
I'll update the Readme to specify minimum version 7.3.

bcholmes commented on webpages/api/admin/module_model.php:
```diff
@@ -99,14 +131,14 @@ public function updateEnabled($db, $value) {
-           ORDER BY M.name, M.id;
-EOD;
+            ORDER BY M.name, M.id;
+        EOD;
```
Same here.

bcholmes reviewed (commented):
Looks good (with a couple minor comments)

bcholmes commented:
> you'd need to add `use PlanZ\Tool;` to your WisCon modules. Would you be okay with that?

Yeah, I can do that.
